pkg/controller/messaginginfra/router/controller.go

```diff
@@ -0,0 +1,188 @@
+
+	return common.WithConditionUpdate(routersCreated, func() error {
+		_, err := controllerutil.CreateOrUpdate(ctx, r.client, statefulset, func() error {
+			if err := controllerutil.SetControllerReference(infra, statefulset, r.scheme); err != nil {
+				return err
+			}
+
+			install.ApplyStatefulSetDefaults(statefulset, "router", infra.Name)
+
+			applyScalingStrategy(infra.Spec.Router.ScalingStrategy, statefulset)
+
+			containers, err := install.ApplyContainerWithError(statefulset.Spec.Template.Spec.Containers, "router", func(container *corev1.Container) error {
+				err := install.ApplyContainerImage(container, "router", infra.Spec.Router.Image)
+				if err != nil {
+					return err
+				}
+
+				install.ApplyVolumeMountSimple(container, "certs", "/etc/enmasse-certs", false)
+				install.ApplyVolumeMountSimple(container, "config", "/etc/qpid-dispatch/config", false)
+
+				install.ApplyEnvSimple(container, "INFRA_NAME", infra.Name)
+				install.ApplyEnvSimple(container, "QDROUTERD_CONF", "/etc/qpid-dispatch/config/qdrouterd.json")
+				install.ApplyEnvSimple(container, "QDROUTERD_CONF_TYPE", "json")
+				install.ApplyEnvSimple(container, "QDROUTERD_AUTO_MESH_DISCOVERY", "INFER")
+				install.ApplyEnvSimple(container, "QDROUTERD_AUTO_MESH_SERVICE_NAME", routerInfraName)
+
+				container.Ports = []corev1.ContainerPort{{
+					ContainerPort: 55672,
+					Name:          "inter-router",
+				}}
+
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+			statefulset.Spec.Template.Spec.Containers = containers
+			statefulset.Spec.ServiceName = routerInfraName
+
+			install.ApplyConfigMapVolume(&statefulset.Spec.Template.Spec, "config", routerInfraName)
+			install.ApplySecretVolume(&statefulset.Spec.Template.Spec, "certs", certSecretName)
+			return nil
+		})
+
+		if err != nil {
+			return err
+		}
+
+		service := &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{Namespace: infra.Namespace, Name: routerInfraName},
+		}
+
+		_, err = controllerutil.CreateOrUpdate(ctx, r.client, service, func() error {
+			if err := controllerutil.SetControllerReference(infra, service, r.scheme); err != nil {
+				return err
+			}
+			install.ApplyServiceDefaults(service, "router", infra.Name)
+			service.Spec.ClusterIP = "None"
+			service.Spec.Selector = statefulset.Spec.Template.Labels
+			service.Spec.Ports = []corev1.ServicePort{
+				{
+					Port:       55672,
+					Protocol:   corev1.ProtocolTCP,
+					TargetPort: intstr.FromString("inter-router"),
+					Name:       "inter-router",
+				},
+			}
+
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		routerConfig := generateConfig(&infra.Spec.Router)
+		routerConfigBytes, err := serializeConfig(&routerConfig)
+		if err != nil {
+			return err
+		}
+		config := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Namespace: infra.Namespace, Name: routerInfraName},
+			Data: map[string]string{
+				"qdrouterd.json": string(routerConfigBytes),
```

Comment:
I think that when you change the config here, the pods of the stateful set will not pick up the changed configuration. For IoT, I've been using the functionality from the `cchange` package to generate a hash, putting that into the deployments.

Reply:
Good point, will fix that.